File: branches/objectweb/TraceAndPerf/performancelibrary/test/test1/java/src/cdmw/performance/test/ClientThread.java
/* =========================================================================== *
 * This file is part of CARDAMOM (R) which is jointly developed by THALES
 * and SELEX-SI.
 * 
 * It is derivative work based on PERCO Copyright (C) THALES 2000-2003.
 * All rights reserved.
 * 
 * CARDAMOM is free software; you can redistribute it and/or modify it under
 * the terms of the GNU Library General Public License as published by the
 * Free Software Foundation; either version 2 of the License, or (at your
 * option) any later version.
 * 
 * CARDAMOM is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
 * FITNESS FOR A PARTICULAR PURPOSE. See the GNU Library General Public
 * License for more details.
 * 
 * You should have received a copy of the GNU Library General
 * Public License along with CARDAMOM; see the file COPYING. If not, write to
 * the Free Software Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
 * =========================================================================== */


package cdmw.performance.test;

import java.io.*;


import com.thalesgroup.CdmwPerformanceAdmin.PerformanceAdmin;
import com.thalesgroup.CdmwPerformanceAdmin.PerformanceAdminHelper;

public class ClientThread extends Thread {

    static final String PERF_ADMIN_IOR_FILE = "perfoLibrary.ior";

    private static final String INFO_MESSAGE =
        "Type h or help for a list of available commands.";
    private static final String PROMPT_MESSAGE = "\n%>";
    private static final String UNKNOWN_COMMAND_MESSAGE = "Bad command!>";

    // Definition of a basic operation.
    // Each operation must redefine the method execute.
    abstract class OperationBase {
        public static final int SUCCESS = 0;
        public static final int FAILURE = -1;
        public static final int EXIT_ORDERED = -10;

        public abstract int execute(String[] arg, java.io.PrintStream os);
    }

    // List of all operation.
    // Key is operation name; an operation could be present
    // several time with different operation name
    private java.util.Hashtable operationsList;

    // ORB
    private org.omg.CORBA.ORB orb;

    // verbose mode
    private boolean verbose;

    // Input stream
    java.io.InputStream istream;

    // Output stream
    java.io.PrintStream ostream;

    public ClientThread(
        org.omg.CORBA.ORB orb,
        java.io.InputStream is,
        java.io.OutputStream os) {
        super("ClientThread");
        this.orb = orb;
        this.verbose = false;
        this.istream = is;
        this.ostream = new java.io.PrintStream(os);

        registerOperations();
    }

    // register definition of available operations
    private void registerOperations() {
        operationsList = new java.util.Hashtable();

        // -- no operation
        OperationBase noOperation = new OperationBase() {
            public int execute(String[] arg, java.io.PrintStream os) {
                return SUCCESS;
            }
        };
        operationsList.put("", noOperation);

        // -- help
        OperationBase printHelp = new OperationBase() {
            public int execute(String[] arg, java.io.PrintStream os) {
                os.println(
                    " +======================+============================================+\n"
                        + " | COMMAND              | DESCRIPTION                                |\n"
                        + " +======================+============================================+\n"
                        + " | x | exit             | exit ClientThread mode                     |\n"
                        + " +----------------------+--------------------------------------------+\n"
                        + " | h | help             | print available commands                   |\n"
                        + " +----------------------+--------------------------------------------+\n"
                        + " | verbose              | toggle verbose mode on/off                 |\n"
                        + " +----------------------+--------------------------------------------+\n"
                        + " | activate <T/F>       | activate/deactivate the measure            |\n"
                        + " +----------------------+--------------------------------------------+\n"
                        + " | get_api              | get info of API's                          |\n"
                        + " +----------------------+--------------------------------------------+\n"
                        + " | start                | start API to measure                       |\n"
                        + " +----------------------+--------------------------------------------+\n");

                return SUCCESS;
            }
        };
        operationsList.put("help", printHelp);
        operationsList.put("h", printHelp);

        // ------- exit -------
        OperationBase exit = new OperationBase() {
            public int execute(String[] arg, java.io.PrintStream os) {
                try {
                    if (verbose)
                        os.println("Initiating ORB shutdown.");
                    orb.shutdown(false);
                } catch (org.omg.CORBA.SystemException ex) {
                    printException(ex, os);
                }
                return EXIT_ORDERED;
            }
        };
        operationsList.put("exit", exit);
        operationsList.put("x", exit);

        // ------- verbose -------
        OperationBase toggleVerbose = new OperationBase() {
            public int execute(String[] arg, java.io.PrintStream os) {
                verbose = !verbose;
                os.println(
                    "(verbose mode is " + (verbose ? "on" : "off") + ")");
                return SUCCESS;
            }
        };
        operationsList.put("verbose", toggleVerbose);

        // ------- activate -------
        OperationBase activatePerfo = new OperationBase() {
            public int execute(String[] arg, java.io.PrintStream os) {
                if (arg.length != 1) {
                    os.println("Error: syntax is " + "'activate <T/F>'");
                    return FAILURE;
                }
 String fileName = PERF_ADMIN_IOR_FILE;
org.omg.CORBA.Object obj=null;
try{
               
                // import the object reference from the file
              /*  org.omg.CORBA.Object obj =
                    orb.string_to_object("file:/" + fileName);*/
                File f = new File(fileName);
                BufferedReader br =
                new BufferedReader( new FileReader( f ));

            // get object reference from command-line argument file
            obj = 
                orb.string_to_object( br.readLine() );

            br.close();
            }catch( Exception ex )  {}
                if (obj == null) {
                    os.println("Error opening reference filename :" + fileName);
                }

                // get the PerformanceAdmin proxy
                PerformanceAdmin perfoAdmin =
                    PerformanceAdminHelper.narrow(obj);

                if (perfoAdmin == null) {
                    ostream.println(
                        "Could not narrow object to type PerformanceAdmin");
                    return FAILURE;
                }

                if (arg[0].compareToIgnoreCase("T") == 0) {
                    perfoAdmin.activate();
                } else if (arg[0].compareToIgnoreCase("F") == 0) {
                    perfoAdmin.deactivate();
                } else {
                    os.println("Error: syntax is " + "'activate <T/F>'");
                    return FAILURE;
                }

                return SUCCESS;
            }
        };
        operationsList.put("activate", activatePerfo);

        // ------- get_api -------
        OperationBase getApiInfo = new OperationBase() {
            public int execute(String[] arg, java.io.PrintStream os) {
                TargetThread target = new TargetThread();

                int nbOfApi = target.getApiNbr();
                for (int i = 0; i < nbOfApi; i++) {
                    String apiName = target.getApiName(i);
                    if (apiName != null) {
                        os.println(
                            "api name          : "
                                + apiName
                                + "\n"
                                + "  sleep time msec : "
                                + target.getApiSleepTime(i)
                                + "\n"
                                + "  call counter    : "
                                + target.getApiCount(i)
                                + "\n");
                    }
                }

                return SUCCESS;
            }
        };
        operationsList.put("get_api", getApiInfo);

        // ------- start -------
        OperationBase startTargets = new OperationBase() {
            public int execute(String[] arg, java.io.PrintStream os) {
                int result = FAILURE;

                try {
                    // create 2 target threads
                    TargetThread target1 = new TargetThread();
                    TargetThread target2 = new TargetThread();

                    // start clients
                    target1.start();
                    target2.start();

                    // wait end of targets
                    target1.join();
                    target2.join();

                    result = SUCCESS;
                } catch (InterruptedException e) {
                    e.printStackTrace(new java.io.PrintWriter(os));
                }

                return result;
            }
        };
        operationsList.put("start", startTargets);

    }

    public void run() {
        java.io.BufferedReader inReader =
            new java.io.BufferedReader(new java.io.InputStreamReader(istream));

        printInfo(ostream);

        int operationResult = 0;
        do {
            ostream.print(PROMPT_MESSAGE);
            try {
                // read command
                String command = inReader.readLine();

                // if Ctrl-C readLine() return null
                if (command == null) {
                    break;
                }

                java.util.StringTokenizer st =
                    new java.util.StringTokenizer(command);
                int nbTokens = st.countTokens();

                // if input not empty
                if (nbTokens != 0) {

                    // parse input
                    String commandName = "";
                    String[] commandArgs = new String[nbTokens - 1];
                    try {
                        commandName = st.nextToken();
                        for (int i = 0; i < commandArgs.length; i++) {
                            commandArgs[i] = st.nextToken();
                        }

                    } catch (java.util.NoSuchElementException e) {
                        // should not append
                    }

                    // if verbose print command
                    if (verbose) {
                        ostream.print("@cmd = <" + commandName + "> ");
                        ostream.print("@args = <");
                        for (int i = 0; i < commandArgs.length; i++) {
                            ostream.print(commandArgs[i]);
                            if (i < commandArgs.length - 1)
                                ostream.print(", ");
                        }
                        ostream.println(">");
                    }

                    // execute command
                    OperationBase op =
                        (OperationBase) operationsList.get(commandName);
                    if (op != null) {
                        operationResult = op.execute(commandArgs, ostream);
                    } else {
                        ostream.println(UNKNOWN_COMMAND_MESSAGE);
                        operationResult = OperationBase.FAILURE;
                    }
                }

            } catch (java.io.IOException e) {
                ostream.println("ERROR in readLine:");
                e.printStackTrace(new java.io.PrintWriter(ostream));
                operationResult = OperationBase.FAILURE;
            }

        } while (operationResult != OperationBase.EXIT_ORDERED);
    }

    private void printInfo(java.io.PrintStream os) {
        os.println(
            "\n\n"
                + "\t\t---------------------------------------\n"
                + "\t\t  Cdmw Performance Test               \n"
                + "\t\t---------------------------------------\n"
                + "\n\n"
                + INFO_MESSAGE
                + "\nInfo: (verbose mode is "
                + (verbose ? "on" : "off")
                + ")");
    }

    private void printException(
        cdmw.common.Exception ex,
        java.io.PrintStream os) {
        os.println("CDMW Exception : ");
        ex.printStackTrace(new java.io.PrintWriter(ostream));
    }

    private void printException(
        org.omg.CORBA.SystemException ex,
        java.io.PrintStream os) {
        os.println("CORBA system Exception : ");
        ex.printStackTrace(new java.io.PrintWriter(ostream));
    }

    private void printException(
        org.omg.CORBA.UserException ex,
        java.io.PrintStream os) {
        os.println("CORBA user Exception : ");
        ex.printStackTrace(new java.io.PrintWriter(ostream));
    }

}

File: tags/V8_0s01/TraceAndPerf/performancelibrary/test/test1/c++/src/ClientThread.cpp
/* ===================================================================== */
/*
 * This file is part of CARDAMOM (R) which is jointly developed by THALES
 * and SELEX-SI. It is derivative work based on PERCO Copyright (C) THALES
 * 2000-2003. All rights reserved.
 * 
 * Copyright (C) THALES 2004-2005. All rights reserved
 * 
 * CARDAMOM is free software; you can redistribute it and/or modify it
 * under the terms of the GNU Library General Public License as published
 * by the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 * 
 * CARDAMOM is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
 * FITNESS FOR A PARTICULAR PURPOSE. See the GNU Library General Public
 * License for more details.
 * 
 * You should have received a copy of the GNU Library General Public
 * License along with CARDAMOM; see the file COPYING. If not, write to the
 * Free Software Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
*/
/* ===================================================================== */


#include "test1/ClientThread.hpp"
#include <string>
#include <cctype>
#include <algorithm>

#include "Foundation/common/System.hpp"
#include "Foundation/ossupport/OS.hpp"
#include "Foundation/orbsupport/OrbSupport.hpp"
#include "Foundation/common/Exception.hpp"
#include "Foundation/common/Assert.hpp"

#include "TraceAndPerf/idllib/CdmwPerformanceAdmin.stub.hpp"
#include "performancelibrary/ProbeCollector.hpp"
#include "performancelibrary/FunctionStat.hpp"

#include "test1/TargetThread.hpp"

#include <sstream>
#include <iostream>
#include <fstream>

#define PRINT_EXCEPTION(f,l,e,os) \
do { \
    os << "File : " << f << " Line : " << l << "\n";\
    print_exception(e,os);\
    CDMW_ASSERT(0); \
} while(0) \

namespace
{
    const char* INFO_MESSAGE = "Type h or help for a list of available commands.";
    const char* PROMPT_MESSAGE = "\n%>";
    const char* UNKNOWN_COMMAND_MESSAGE = "Bad command!>";

    const CORBA::Long EXIT_ORDERED(-10);
    const CORBA::Long OP_SUCCESS = 0;
    const CORBA::Long OP_FAILURE = -1;
       

    void print_exception(const CORBA::SystemException & ex, std::ostream & os)
    {
        os << "CORBA system Exception : \n" << ex << std::endl;
    }

    void print_exception(const CORBA::Exception & ex, std::ostream & os)
    {
        os << "CORBA Exception : \n" << ex << std::endl;
    }

	void print_exception(const Cdmw::Exception & ex, std::ostream & os)
	{
	    os << "CDMW Exception : \n" << ex.what() << std::endl;
	}

    
    bool not_space(char c)
    { return !isspace(c);}

}; // End anonymous namespace

namespace Cdmw
{
namespace Performance
{

//----------------------------------------------------------------------------
ClientThread::ClientThread(CORBA::ORB_ptr orb, 
                           std::istream & is,
                           std::ostream & os)
    throw (CORBA::SystemException)
    : m_orb(CORBA::ORB::_duplicate(orb)),
      m_verbose(false),
      m_istream(is),
      m_ostream(os)
{
}

//----------------------------------------------------------------------------
ClientThread::~ClientThread() throw()
{
}

//----------------------------------------------------------------------------
void ClientThread::run() throw()
{
    run (m_istream,m_ostream);
}

//----------------------------------------------------------------------------
const CORBA::ULong ClientThread::nb_commands = 10;
const ClientThread::command_def ClientThread::commands[] =  // should be sorted alphabetically (by operation name)
{ 
    { "activate"          , &ClientThread::activate_perfo              },  
    { "exit"              , &ClientThread::exit                        },
	{ "get_api"           , &ClientThread::get_api_info                },
	{ "get_stat"          , &ClientThread::get_function_stat           },
    { "h"                 , &ClientThread::print_help                  },
    { "help"              , &ClientThread::print_help                  },
	{ "start"             , &ClientThread::start_targets               },
    { "verbose"           , &ClientThread::toggle_verbose              },
	{ "write_stat"        , &ClientThread::write_stat                  },
    { "x"                 , &ClientThread::exit                        }
};

    
//----------------------------------------------------------------------------
CORBA::Long ClientThread::runOperation(const std::string &op,
                                       const std::string &arg, 
                                       std::ostream &os) throw()
{
    CORBA::Long l = 0;
    CORBA::Long r = (CORBA::Long)nb_commands;
    CORBA::Long m;
    CORBA::Long res;
    
    while(l < r)
    {
      m = (l + r) / 2;
      res = strcmp(commands[m].name, op.c_str());
	  
      if(res == 0)
      {
	    // run command
        return (this->*commands[m].op)(arg,os);
	  }
      
	  if(res > 0)
	  {
        r = m;
	  }
      else
	  {
        l = m + 1;
	  }
    }
	
    if(op != "") unknown_command(os);
    
    return OP_FAILURE;
}

//----------------------------------------------------------------------------
void ClientThread::print_info(std::ostream& os) const throw()
{
    os << "\n\n"
       << "\t\t---------------------------------------\n"
       << "\t\t  Cdmw Performance Test               \n"
       << "\t\t---------------------------------------\n"
       << "\n\n"
       << INFO_MESSAGE
       << "\nInfo: (verbose mode is " << (m_verbose ? "on" : "off") << ")"
       << std::endl;
}

//----------------------------------------------------------------------------
void ClientThread::print_prompt(std::ostream& os) const throw()
{
    os << PROMPT_MESSAGE << std::flush;
}

//----------------------------------------------------------------------------
void ClientThread::unknown_command( std::ostream & os) const throw()
{
    os << UNKNOWN_COMMAND_MESSAGE << std::endl;
}

//----------------------------------------------------------------------------
void ClientThread::run(std::istream& is,std::ostream& os) throw()
{
    print_info(os);
    print_prompt(os);
	
    std::string s;
	
    while(std::getline(is,s)) 
	{
        // ===============
        // Extract command        
        // ===============
        // skip leading whitespaces
        std::string::const_iterator i = s.begin();
        std::string::const_iterator end = s.end();
        i = std::find_if(i,end,not_space);
        // find end of command
        std::string::const_iterator j = std::find_if(i,end,isspace);
        // copy the characters in [i,j)
        std::string command;
        if (i != end)
            command = std::string(i,j);
		
        // =================
        // Extract arguments
        // =================
        // skip leading whitespaces
        i = std::find_if(j,end,not_space);
        std::string arg;
        // copy arguments
        if (i != end)
            arg = std::string(i,end);
        
        if (runOperation(command,arg,os) == EXIT_ORDERED) break; 
        print_prompt(os); 
    } 
}

//----------------------------------------------------------------------------
CORBA::Long ClientThread::exit(const std::string & arg, std::ostream & os) 
    throw()
{    
    try 
   	{
        if (m_verbose) os << "Initiating ORB shutdown." << std::endl;
        m_orb->shutdown(false);
    } 
	catch (const CORBA::SystemException & ex) 
	{
        PRINT_EXCEPTION(__FILE__,__LINE__,ex,os);
    }
    return EXIT_ORDERED;
}

//----------------------------------------------------------------------------
CORBA::Long ClientThread::toggle_verbose(const std::string & arg, std::ostream & os) 
    throw()
{    
    m_verbose = !m_verbose;
    os << "(verbose mode is " << (m_verbose ? "on" : "off") << ")"
       << std::endl;
    return OP_SUCCESS;
}

//----------------------------------------------------------------------------
// Available commands:
//
CORBA::Long ClientThread::print_help(const std::string & arg, std::ostream & os)
    throw()
{
     os << " +======================+============================================+" << "\n"
        << " | COMMAND              | DESCRIPTION                                |" << "\n"
        << " +======================+============================================+" << "\n"
        << " | x | exit             | exit ClientThread mode                     |" << "\n"
        << " +----------------------+--------------------------------------------+" << "\n"
        << " | h | help             | print available commands                   |" << "\n"
        << " +----------------------+--------------------------------------------+" << "\n" 
        << " | verbose              | toggle verbose mode on/off                 |" << "\n"
        << " +----------------------+--------------------------------------------+" << "\n"
		<< " | activate <T/F>       | activate/deactivate the measure            |" << "\n"
		<< " +----------------------+--------------------------------------------+" << "\n"
		<< " | get_api              | get info of API's                          |" << "\n"
	    << " +----------------------+--------------------------------------------+" << "\n"
		<< " | get_stat <API name>  | get statistics for the api name            |" << "\n"
		<< " +----------------------+--------------------------------------------+" << "\n"
	    << " | start                | start API to measure                       |" << "\n"
	    << " +----------------------+--------------------------------------------+" << "\n"
		<< " | write_stat <file>    | write statistics in specified file         |" << "\n"
	    << " +----------------------+--------------------------------------------+" << "\n"
				
        << std::endl;

     return OP_SUCCESS;
}

//----------------------------------------------------------------------------
CORBA::Long ClientThread::activate_perfo(const std::string &arg, std::ostream &os) throw()
{
  CORBA::Long result = OP_FAILURE;
  const char *p_arg = arg.c_str();
  
  try 
  {
    CORBA::Object_var objref;
	char *p_file_name = "perfoLibrary.ior";

	// import the object reference from the file
	std::ifstream is (p_file_name);
	if (is.good())
	{
	  char objref_text[5000];
	  is >> objref_text;
	  objref = m_orb->string_to_object(objref_text);

	  is.close();
	}
	else
	{
      os << "File : " << __FILE__ << " Line : " << __LINE__
	     << "error opening reference filename :" << p_file_name << std::endl;
	  return result;
	}
	
	// get the PerformanceAdmin proxy
	CdmwPerformanceAdmin::PerformanceAdmin_var perfoAdmin = 
			     CdmwPerformanceAdmin::PerformanceAdmin::_narrow(objref.in());

	if (CORBA::is_nil(perfoAdmin.in()))
	{
	  os << "File : " << __FILE__ << " Line : " << __LINE__
	     << "could not _narrow object to type PerformanceAdmin" << std::endl;
	  return result;
	}

	// activate or deactivate the performance admin according parameter
	if (*p_arg == 'T' || *p_arg == 't')
	{
	  perfoAdmin->activate();
	}
	else
	{
	  perfoAdmin->deactivate();
	}

	result = OP_SUCCESS;
  } 
  
  catch (const Cdmw::Exception &e ) 
  {
	PRINT_EXCEPTION(__FILE__,__LINE__,e,os);
	
  }
  catch (const CORBA::SystemException &e)
  {
	PRINT_EXCEPTION(__FILE__,__LINE__,e,os);
  }

  return result;
}

//----------------------------------------------------------------------------
CORBA::Long ClientThread::start_targets(const std::string &arg, std::ostream &os) throw()
{
  CORBA::Long result = OP_FAILURE;
   
  try
  {
    // create 2 target threads
    Performance::TargetThread target1;
    Performance::TargetThread target2;
  
    // start clients
    target1.start();
    target2.start();

    // wait end of targets
    target1.join();
    target2.join();

    result = OP_SUCCESS;
  } 

  catch (const Cdmw::Exception & e )
  {
	PRINT_EXCEPTION(__FILE__,__LINE__,e,os);	
  }
	
  return result;
}

//----------------------------------------------------------------------------
CORBA::Long ClientThread::get_api_info(const std::string &arg, std::ostream &os) throw()
{
  TargetThread target;

  int nbr_of_api = target.getApiNbr();
  std::string api_name; 
  int i;

  for (i=0 ; i < nbr_of_api ; i++)
  {
    target.getApiName (i, api_name);
    os << "api name          : " << api_name.c_str() << std::endl; 
	os << "  sleep time msec : " << target.getApiSleepTime(i) << std::endl;	
	os << "  call counter    : " << target.getApiCount(i) << std::endl;	
	os << std::endl;
  }

  return OP_SUCCESS;
}

//----------------------------------------------------------------------------
CORBA::Long ClientThread::get_function_stat(const std::string &arg, std::ostream &os) throw()
{
  CORBA::Long result = OP_FAILURE;
  const char *p_arg = arg.c_str();
     
  try
  {					  
    Performance::ProbeCollector *p_collector = NULL;

    // get the Probe Collector
    p_collector = Performance::ProbeCollector::GetInstance();
  
    // get the Function stat	
    if (p_arg != NULL)
	{
	  const Performance::FunctionStat *p_function_stat = 
			                                  p_collector->getFunctionStat(arg);

	  if (p_function_stat != NULL)
	  {
        os << "api name : " << p_function_stat->getFunctionName().c_str()
		   << std::endl
		   << "min elapsed time (microsec) : " << p_function_stat->getMinElapsedTime()
		   << std::endl
		   << "max elapsed time (microsec) : " << p_function_stat->getMaxElapsedTime()
		   << std::endl
		   << "sum elapsed time (microsec) : " << p_function_stat->getSumElapsedTime()
		   << std::endl
		   << "hit count : " << p_function_stat->getHitCount()
	       << std::endl;
      }
      else
	  {
	     os << "api name is not recorded in collector" << std::endl;
	  }	  
	}
	else
	{
      os << "api name is missing" << std::endl;
	}
	
   	result = OP_SUCCESS;
  } 

  catch (const Cdmw::Exception & e )
  {
	PRINT_EXCEPTION(__FILE__,__LINE__,e,os);	
  }
	
  return result;
}

//----------------------------------------------------------------------------
CORBA::Long ClientThread::write_stat(const std::string & arg, std::ostream & os) throw()
{
  CORBA::Long result = OP_FAILURE;
  const char *p_arg = arg.c_str();
     
  try
  {					  
    Performance::ProbeCollector *p_collector = NULL;

    // get the Probe Collector
    p_collector = Performance::ProbeCollector::GetInstance();
  
    // write statistic in file passed as argument
    if (p_arg != NULL)
	{
	  p_collector->writeStatistic(arg);
	}
	else
	{
      os << "file name is missing" << std::endl;
	}
	
   	result = OP_SUCCESS;
  } 

  catch (const Cdmw::Exception & e )
  {
	PRINT_EXCEPTION(__FILE__,__LINE__,e,os);	
  }
	
  return result;
}	
				
//----------------------------------------------------------------------------

}; // namespace Performance
}; // namespace Cdmw

